File: codes/semana12/ex01_Notificacao.py
from abc import ABC, abstractmethod

# Produto abstrato
class Notificacao(ABC):

    @abstractmethod
    def enviar(self, mensagem):
        pass


# Produtos concretos
class Email(Notificacao):

    def enviar(self, mensagem):
        print(f"[EMAIL] {mensagem}")


class SMS(Notificacao):

    def enviar(self, mensagem):
        print(f"[SMS] {mensagem}")


class Push(Notificacao):

    def enviar(self, mensagem):
        print(f"[PUSH] {mensagem}")


# Factory Method
class FabricaNotificacao:

    @staticmethod
    def criar_notificacao(tipo):

        if tipo == "email":
            return Email()

        elif tipo == "sms":
            return SMS()

        elif tipo == "push":
            return Push()

        else:
            raise ValueError("Tipo inválido")


# Programa principal
tipo = input("Digite o tipo (email/sms/push): ")

notificacao = FabricaNotificacao.criar_notificacao(tipo)

notificacao.enviar("Bem-vindo ao sistema!")

File: codes/semana12/Ex01_Notificacao.cpp
#include <iostream>
#include <memory>

using namespace std;

// Produto abstrato
class Notificacao {
public:
    virtual void enviar(string mensagem) = 0;
    virtual ~Notificacao() {}
};

// Produtos concretos
class Email : public Notificacao {
public:
    void enviar(string mensagem) override {
        cout << "[EMAIL] " << mensagem << endl;
    }
};

class SMS : public Notificacao {
public:
    void enviar(string mensagem) override {
        cout << "[SMS] " << mensagem << endl;
    }
};

class Push : public Notificacao {
public:
    void enviar(string mensagem) override {
        cout << "[PUSH] " << mensagem << endl;
    }
};

// Factory
class FabricaNotificacao {
public:

    static unique_ptr<Notificacao> criar(string tipo) {

        if (tipo == "email")
            return make_unique<Email>();

        else if (tipo == "sms")
            return make_unique<SMS>();

        else if (tipo == "push")
            return make_unique<Push>();

        return nullptr;
    }
};

int main() {

    string tipo;

    cout << "Digite o tipo (email/sms/push): ";
    cin >> tipo;

    auto notificacao = FabricaNotificacao::criar(tipo);

    if (notificacao)
        notificacao->enviar("Bem-vindo ao sistema!");
    else
        cout << "Tipo invalido!" << endl;

    return 0;
}